### DISCUSSION TITLE
[WebGPU] device.importExternalTexture and queue.copyExternalImageToTexture are missing

### DISCUSSION THREAD

**Discussion by alienself** [emscripten-core/emscripten#18946] [Q&A] [answered]
Hi,

I would like to recreate this [example](https://austin-eng.com/webgpu-samples/samples/texturedCube) and [this one](https://austin-eng.com/webgpu-samples/samples/videoUploading) using c++ and emscripten however the api call for device.importExternalTexture and queue.copyExternalImageToTexture are completely missing from the binding class.

This means that we currently cannot load up external images nor video texture within emscripten + webgpu :(

how should I proceed? 

https://austin-eng.com/webgpu-samples/samples/texturedCube
https://austin-eng.com/webgpu-samples/samples/videoUploading

**Comment by kripken** [comment]
cc @kainino0x

**Comment by kainino0x** [comment, marked as answer]
This is an open item, but I'm not exactly sure what the fix is. These APIs require DOM objects to be passed in (which live outside the WASM heap), so need to take something like an `EM_VAL`. But I don't know what the appropriate Emscripten-y way to do this is.

For now IIRC the workaround is to use `EM_ASM` to make these calls. It takes a little effort to get the correct objects into JS to be able to do this but it should be possible.

**Reply by alienself** [reply]
Do you know if someone already wrote something like this? 
How can access the webgpu handles created by emscripten through javascript?
